[analyzer] Harden RegionStoreManager::bindArray

marco-antognini-sonarsource opened the pull request:
Fixes https://github.com/llvm/llvm-project/issues/147686 by handling symbolic values similarly to bindStruct and handling constant values. The latter is actually more of a workaround: bindArray should not have to deal with such constants.

CPP-6688

marco-antognini-sonarsource committed:
[analyzer] Harden RegionStoreManager::bindArray

Fixes https://github.com/llvm/llvm-project/issues/147686 by completing
handling of literals and handling symbolic values similarly to
bindStruct.

CPP-6688

NagyDonat commented on clang/lib/StaticAnalyzer/Core/RegionStore.cpp:
```diff
@@ -2654,14 +2654,19 @@ RegionStoreManager::bindArray(LimitedRegionBindingsConstRef B,
     return bindAggregate(B, R, V);
   }
+  if (auto const *Value = Init.getAsInteger()) {
+    auto SafeValue = StateMgr.getBasicVals().getValue(*Value);
```
Perhaps add a source code comment to explain why you need to unwrap and reconstruct this value (instead of e.g. just returning `bindAggregate(B, R, Init)`.

NagyDonat commented on clang/lib/StaticAnalyzer/Core/RegionStore.cpp:
```diff
@@ -2654,14 +2654,19 @@ RegionStoreManager::bindArray(LimitedRegionBindingsConstRef B,
+  }
 
-  // Handle lazy compound values.
+  // Handle lazy compound values and symbolic values.
```
Why are you highlighting lazy compound values and symbolic values together here? Is there something common between them? If not, then I'd prefer removing this comment (instead of extending it) because it doesn't say anything that isn't immediately visible from the code.

NagyDonat commented on clang/test/Analysis/initializer.cpp:
```diff
@@ -610,3 +610,44 @@ void top() {
+  // FIXME The next eval should result in TRUE.
+  clang_analyzer_eval(*a.m_ptr == 0); // expected-warning{{UNKNOWN}}
+}
+
+void test2() {
+  A a(314);
+  clang_analyzer_eval(a.m_buf[0] == 0); // expected-warning{{TRUE}}
+  clang_analyzer_eval(*a.m_ptr == 0); // expected-warning{{TRUE}}
+}
+
+void test3() {
+  A a(0);
+  clang_analyzer_eval(a.m_buf[0] == 0); // expected-warning{{TRUE}}
+  clang_analyzer_eval(*a.m_ptr == 0); // expected-warning{{TRUE}}
+}
+
+void test4() {
+  A a(0, 314);
+  clang_analyzer_eval(a.m_buf[0] == 0); // expected-warning{{TRUE}}
+  clang_analyzer_eval(*a.m_ptr == 0); // expected-warning{{TRUE}}
+}
```
```suggestion
void test3(char arg) {
  A a(arg);
  clang_analyzer_eval(a.m_buf[0] == arg); // expected-warning{{TRUE}}
  clang_analyzer_eval(*a.m_ptr == arg); // expected-warning{{TRUE}}
}

void test4(char arg) {
  A a(arg, 314);
  clang_analyzer_eval(a.m_buf[0] == arg); // expected-warning{{TRUE}}
  clang_analyzer_eval(*a.m_ptr == arg); // expected-warning{{TRUE}}
}
```
To me it seems that these tests tried to test the branch `if (isa<nonloc::SymbolVal>(Init))` but instead of using truly symbolic value they just use ConcreteInt zero in a slightly more complex way.

(Please check that the tests actually pass with my suggested change -- although the analyzer should be smart enough to realize that a symbol is equal to itself.)

NagyDonat replied:
Ok, thanks for updating the tests. 

>  I won't have time to investigate though.

No problem, the unexpected behavior of `test3Bis` is unrelated to the bug fixed in this commit (it seems to be a limitation of our modeling of the intricate details of C++ variable initialization). Just adding the FIXME is perfectly fine.

NagyDonat reviewed (commented):
Thanks for the patch! I have some minor remarks inline but overall I like the patch and I completely agree with the maxim that "not crashing is strictly better than crashing".

Also thanks @steakhal for clarifying that this is a "hotfix" and covers situations that "do not belong to" `bindArray`. I agree that this should be mentioned in the commit message; in fact I think that we should also mention this in a `FIXME` source code comment, because it would be really valuable for those who try to understand the purpose of this method (_It is called bindArray... but it also handles integer literals... maybe it's misnamed... ?_).

marco-antognini-sonarsource committed:
Merge remote-tracking branch 'upstream/main' into mb/bindArray

marco-antognini-sonarsource committed:
Address feedback: spell out type

marco-antognini-sonarsource committed:
Address feedback: remove dubious comment

marco-antognini-sonarsource committed:
Address feedback: use proper symbolic value for test4

marco-antognini-sonarsource committed:
Address feedback: document current inconsistency

marco-antognini-sonarsource committed:
Address feedback: simplify bindArray & add FIXME

No need to unwrap constants.
This is a workaround to not crash -- not a proper fix.

marco-antognini-sonarsource reviewed (commented):
@NagyDonat I appreciate your patience on this PR. I've managed to find time to address your feedback. Let me know if you have more of course.

NagyDonat reviewed (approved):
LGTM, thanks for the updates!

It's a bit unfortunate that `test3Bis` doesn't show the ideal behavior, but this commit is a clear step forward and there is no expectation to fix everything with one commit.
